File: Solutions/MoNaDDu/Game Shop/2/Part2.cpp
#include <iostream>
#include <string>
#include <iomanip>
#include <vector>
#include <memory>

using namespace std;

class Item
{
public:
	Item(const string& name, const string& description, const int weight, const int value)
		: name(name), description(description), weight(weight), value(value)
	{

	}

	virtual ~Item() = default;

	virtual void Describe(void) const
	{
		DescribeName();
		DescribeDescription();
		DescribeWeight();
		DescribeValue();
	}

protected:
	static int GetSpaceInfo(void)
	{
		return space;
	}

private:
	string name;
	string description;
	int weight;
	int value;
	static const int space = 12;

	inline void DescribeName(void) const
	{
		cout << left << setw(space) << "Name" << "= " << name << endl;
	}

	inline void DescribeDescription(void) const
	{
		cout << left << setw(space) << "Description" << "= " << description << endl;
	}

	inline void DescribeWeight(void) const
	{
		cout << left << setw(space) << "Weight" << "= " << weight << (weight > 1 ? " lbs" : " lb") << endl;
	}

	inline void DescribeValue(void) const
	{
		cout << left << setw(space) << "Value" << "= " << value << (value > 1 ? " gold coin" : " gold coins") << endl;
	}
};

class Weapon : public Item
{
public:
	Weapon(const string& name, const string& description, const int weight, const int value, const int damage)
		: Item(name, description, weight, value), damage(damage)
	{

	}

	virtual ~Weapon() = default;

	virtual void Describe(void) const override
	{
		Item::Describe();
		DescribeDamage();
	}

private:
	int damage;

	inline void DescribeDamage(void) const
	{
		cout << setw(GetSpaceInfo()) << "Damage" << "= " << damage << endl;
	}
};

class Armor : public Item
{
public:
	Armor(const string& name, const string& description, const int weight, const int value, const int defense)
		: Item(name, description, weight, value), defense(defense)
	{

	}

	virtual ~Armor() = default;

	virtual void Describe(void) const override
	{
		Item::Describe();
		DescribeDefense();
	}

private:
	int defense;

	inline void DescribeDefense(void) const
	{
		cout << setw(GetSpaceInfo()) << "Defense" << "= " << defense << endl;
	}
};

int main()
{
	using ItemPtr = shared_ptr<Item>;

	vector<ItemPtr> item_list{
		make_shared<Weapon>("Excalibur", "The legendary sword of King Arthur", 12, 1024, 24),
		make_shared<Armor>("Steel Armor", "Protective covering made by steel", 15, 805, 18)
	};

	for (const auto& item : item_list)
	{
		item->Describe();
		cout << endl;
	}

    return 0;
}

File: Solutions/MoNaDDu/Game Shop/2/Part2.hs
module Part2 where

data Item = Item { name :: String, description :: String, weight :: Int, value :: Int } |
            Weapon { name :: String, description :: String, weight :: Int, value :: Int, damage :: Int } |
            Armor { name :: String, description :: String, weight :: Int, value :: Int, defense :: Int }
 deriving ( Show )

type Items = [Item]

testCase1 = describe items

describe :: Items -> IO ()
describe = mapM_ (putStrLn . toString)

toString :: Item -> String
toString (Item name description weight value) = baseInfo (Item name description weight value)
toString (Weapon name description weight value damage) = baseInfo (Item name description weight value) ++
                                                         "Damage      = " ++ show damage ++ "\n"
toString (Armor name description weight value defense) = baseInfo (Item name description weight value) ++
                                                         "Defense     = " ++ show defense ++ "\n"

baseInfo :: Item -> String
baseInfo item = "Name        = " ++ name item ++ "\n" ++
                "Description = " ++ description item ++ "\n" ++
                "Weight      = " ++ w'' ++ "\n" ++
                "value       = " ++ v'' ++ "\n"
                where 
                w = weight item
                v = value item
                w' = show w
                v' = show v
                w''
                 | w > 1 = w' ++ " lbs"
                 | otherwise = w' ++ " lb"
                v''
                 | v > 1 = v' ++ " gold coins"
                 | otherwise = v' ++ " gold coin"

items = [
    Weapon { name = "Excalibur", description = "The legendary sword of King Arthur", weight = 12, value = 1024, damage = 24 },
    Armor { name = "Steel Armor", description = "Protective covering made by steel", weight = 15, value = 805, defense = 18}
    ]